src/osyris/core/subdomain.py

```diff
@@ -0,0 +1,159 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2022 Osyris contributors (https://github.com/osyris-project/osyris)
+
+import numpy as np
+from .dataset import Dataset
+from .datagroup import Datagroup
+from .array import Array
+from .vector import Vector
+from .. import spatial
+from .. import units
+
+
+class Subdomain(Dataset):
+    def __init__(self,
+                 dataset,
+                 selection,
+                 dr=None,
+                 dx=None,
+                 dy=None,
+                 dz=None,
+                 origin=None,
+                 basis=None,
+                 dr_L=None):
+
+        self._parent = dataset  # store reference to parent dataset
+        self.meta = dataset.meta
+        self.units = dataset.units
+
+        self._sink_particles = False
+        self._particles = False
+
+        if origin is None:
+            self.origin = Vector(x=0, y=0, z=0, unit=dataset.units['length'].units)
+        else:
+            self.origin = origin
+
+        # find subdomain and extract data
+        valid_comp_amr, valid_comp_sinks, valid_comp_particles = self._determine_subdomain(
+            selection, dr, dx, dy, dz)
+        self._extract_data(valid_comp_amr, valid_comp_sinks, valid_comp_particles)
+
+        # translate positions in subdomain with new origin
+        spatial.change_origin(self, self.origin)
+
+        if basis is None:
+            self.basis = [0, 0, 1]  # assume it's the ramses grid
+        elif isinstance(basis, str):
+            if dr_L is None:
+                raise ValueError(
+                    "Please provide the radius size with which to compute angular momentum (dr_L)"
+                )
+            ang_mom = spatial.get_ang_mom(self, dr_L)
+            if basis.lower() == "top":
+                basis = ang_mom / ang_mom.norm
+            elif basis.lower() == "side":
+                perp_v = Vector(1.0,
+                                1.0,
+                                (-1.0 * (ang_mom.x + ang_mom.y) / ang_mom.z).values,
+                                unit=ang_mom.unit)
+                basis = perp_v / perp_v.norm
+            spatial.change_basis(self, basis)
+        else:
+            spatial.change_basis(self, basis)
+
+    def _determine_subdomain(self, selection, dr, dx, dy, dz):
+        # find indices of data within subdomain for amr dependent groups, sinks and particles
+        centered_pos = self._parent["amr"]["position"] - self.origin
+        if "sink" in self._parent:
```

Comment:
I don't really understand why you need to treat the groups on a case-by-case basis. Now that we have standardized the name for the position vector in all groups to `"position"`, it should be easy enough to write something like
```Py
out = Dataset()
for name, group in dataset.items():
    # the `hydro` and friends groups have no position
    pos = group.get("position", group.parent["amr"]["position"])
    inds = (pos - origin).norm <= radius
    out[name] = group[inds]
```

Reply:
Maybe this is my ramses_ism bias, but ramses outputs a sink file even if there are none (eg. when the criterion for sink creation hasn't been reached yet). The result is an empty sink group in the dataset, which causes errors when trying to access the `position` component. This is why I added the `if len(self._parent["sink"]) > 0:` check. So I treated things in a case-by-case basis to avoid having empty sink/particle groups in the subdomain.

Reply:
If the original data has an empty `sink` group, I think it's ok to have one in the subdomain also.

But this wasn't the main point I was getting it. I think storing indices for amr, sink and particles separately in unnecessary. You should just have a simple loop over the group. I don't think you need to store anything in temporaries? This is probably related to my comment below about why you are using two steps to make the subdomain. First you store the indices that are valid for your subdomain, and then you actually perform the selection. I think you can just do it all in one go, as in the code snippet I suggested above.